### ISSUE TITLE
Fix Quarkus Issue with CRD Registration

### ISSUE THREAD

**Issue by csviri** [java-operator-sdk/kubernetes-webooks-framework#28]
There is an issue with de-serialization for conversion hooks with Quarkus:
When a conversion controller is created a based on the mappers the related custom resources are registered using the fabric8 client API:

https://github.com/java-operator-sdk/admission-controller-framework/blob/3a2019cdbcd9783678613b29cb91c066046b458b/core/src/main/java/io/javaoperatorsdk/webhook/conversion/Utils.java#L22-L25

This is called when the controller is created:
https://github.com/java-operator-sdk/admission-controller-framework/blob/3a2019cdbcd9783678613b29cb91c066046b458b/samples/quarkus/src/main/java/io/javaoperatorsdk/webhook/admission/sample/quarkus/conversion/ConversionControllerConfig.java#L16-L20

However this does not seems to be working (/picked up) by Quarkus. To reproduce see tests:

https://github.com/java-operator-sdk/admission-controller-framework/blob/3a2019cdbcd9783678613b29cb91c066046b458b/samples/quarkus/src/test/java/io/javaoperatorsdk/webhook/admission/sample/quarkus/conversion/ConversionEndpointTest.java#L26-L34

which are calling the edpoints:

https://github.com/java-operator-sdk/admission-controller-framework/blob/3a2019cdbcd9783678613b29cb91c066046b458b/samples/quarkus/src/main/java/io/javaoperatorsdk/webhook/admission/sample/quarkus/conversion/ConversionEndpoint.java#L29-L44

The expected behavior would be that in `ConversionReview.ConversionRequest.objects` are de-serialized properly, not into `GenericKubernetesResource`  but into `MultiVersionCustomResource`.

It is not working neither for async and/or sync endpoints. Interesting is if both tests are run, the second one passes. So eventually the configuration is getting picked up.

It works for Spring boot tests:

https://github.com/java-operator-sdk/admission-controller-framework/blob/3a2019cdbcd9783678613b29cb91c066046b458b/samples/spring-boot/src/test/java/io/javaoperatorsdk/webhook/sample/springboot/conversion/ConversionEndpointTest.java#L37-L46

**Comment by geoand** [comment]
Thanks. I'll try and have a look this week

**Comment by geoand** [comment]
The problem here is that the test code relies on an `ObjectMapper` that is likely **not** the same as the one used by the Kubernetes Client.

Essentially the test is posting to a `ConversionReview` to REST Endpoint which implicitly means that the default ObjectMapper of Quarkus comes into play (see [this](https://github.com/quarkusio/quarkus/blob/2.14.1.Final/independent-projects/resteasy-reactive/server/jackson/src/main/java/org/jboss/resteasy/reactive/server/jackson/JacksonBasicMessageBodyReader.java#L49)).

A simpler test that shows the exact same problem would be the following:

```
  @Inject
  ObjectMapper mapper;

  @Test
  void jacksonTest() throws IOException {
    ObjectReader reader = mapper.reader();
    ConversionReview cr = reader.forType(reader.getTypeFactory().constructType(ConversionReview.class))
            .readValue(this.getClass().getResourceAsStream("/conversion-request.json"));
    List<HasMetadata> objects = cr.getRequest().getObjects();
    assertThat(objects).hasSize(2).allSatisfy(h -> assertThat(h).isInstanceOf(MultiVersionCustomResource.class));
  }
```

This test will fail for the exact same reason: Jackson is not configured properly, so the de-serialization results in the the `objects` field not using the proper deserializer.

My _guess_ as to why the test works the second time, is that your code or the k8s client code registers a deserializer with the default `ObjectMapper`.

Long story short: There is nothing wrong with the Quarkus code in this case :)

**Comment by csviri** [comment]
Thing is if I add this it works:

https://github.com/java-operator-sdk/admission-controller-framework/blob/440a0e5f055d647868970cce95837867f3a8893e/samples/quarkus/src/main/java/io/javaoperatorsdk/webhook/admission/sample/quarkus/conversion/CustomResourceDeserializationCustomizer.java#L11-L21

But that calls the same static code that is also called when creating the conversion controller:

https://github.com/java-operator-sdk/admission-controller-framework/blob/440a0e5f055d647868970cce95837867f3a8893e/core/src/main/java/io/javaoperatorsdk/webhook/conversion/ConversionController.java#L29-L30

The Deserializer is set on the class level:

https://github.com/fabric8io/kubernetes-client/blob/f63fa5faf65787ee31841ce5f9d93ccb8af31280/kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/KubernetesResource.java#L22-L24

So according to my (limited knowledge) of jackson, this should be picked up, and it is picked up just not in case the static part is extended in the later way.

**Comment by geoand** [comment]
> Thing is if I add this it works:
> https://github.com/java-operator-sdk/admission-controller-framework/blob/440a0e5f055d647868970cce95837867f3a8893e/samples/quarkus/src/main/java/io/javaoperatorsdk/webhook/admission/sample/quarkus/conversion/CustomResourceDeserializationCustomizer.java#L11-L21

Yeah, but that is the expected behavior. That's the way users are supposed to customize the `ObjectMapper` provided by Quarkus.

> But that calls the same static code that is also called when creating the conversion controller

Sure, but that likely happens to late

**Comment by csviri** [comment]
> Sure, but that likely happens to late

Hmm but its when the service is initialized that is used by controller.

Ok, will probably end up writing a little extension. 

Thank you very much for help! Will close the issue after PR merged.

**Comment by geoand** [comment]
> Thank you very much for help! Will close the issue after PR merged.

Sure thing! Register a  customizer in an extension is definitly the way to go
